sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -712,6 +711,49 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
 }
 
 /**
+  * Reorder the joins so that the bottom ones have at least one condition.
+  */
+object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
+
+  /**
+    * Reorder the joins so that the bottom ones have at least one condition.
+    */
+  def reorder(
+      input: LogicalPlan,
+      joins: Seq[LogicalPlan],
+      conditions: Seq[Expression]): LogicalPlan = {
+    // filter out the conditions that could be pushed down to `joined`
+    val otherConditions = conditions.filterNot { cond =>
+      cond.references.subsetOf(input.outputSet)
+    }
+    if (joins.isEmpty) {
```

Comment:
I would have the pattern return `None` if there are no joins so you don't have to handle this case.